File: apps/web/src/workflow.ts
import type { StudyStatus } from "./types";

export const stages: { status: StudyStatus; label: string }[] = [
  { status: "DRAFT", label: "研究问题" },
  { status: "PROTOCOL_REVIEW", label: "协议审批" },
  { status: "FEASIBILITY_RUNNING", label: "联邦统计" },
  { status: "FEASIBILITY_REVIEW", label: "可行性审查" },
  { status: "CONTRACT_LOCKED", label: "实验合同" },
  { status: "TRAINING_RUNNING", label: "联邦实验" },
  { status: "RESULTS_REVIEW", label: "结果评审" },
  { status: "PRIVACY_REVIEW", label: "隐私复核" },
  { status: "REPORT_READY", label: "研究资产" },
];

export const statusIndex = (status: StudyStatus) => {
  if (status === "FAILED_RETRYABLE" || status === "FAILED_FINAL") {
    return stages.findIndex((stage) => stage.status === "TRAINING_RUNNING");
  }
  return stages.findIndex((stage) => stage.status === status);
};

File: rarelink/services/workflow.py
from rarelink.domain import StudyStatus


class InvalidTransition(ValueError):
    pass


ALLOWED_TRANSITIONS: dict[StudyStatus, set[StudyStatus]] = {
    StudyStatus.DRAFT: {StudyStatus.PROTOCOL_REVIEW},
    StudyStatus.PROTOCOL_REVIEW: {StudyStatus.FEASIBILITY_RUNNING},
    StudyStatus.FEASIBILITY_RUNNING: {
        StudyStatus.FEASIBILITY_REVIEW,
        StudyStatus.BLOCKED_BY_POLICY,
        StudyStatus.FAILED_RETRYABLE,
    },
    StudyStatus.FEASIBILITY_REVIEW: {StudyStatus.CONTRACT_LOCKED},
    StudyStatus.CONTRACT_LOCKED: {StudyStatus.TRAINING_RUNNING},
    StudyStatus.TRAINING_RUNNING: {
        StudyStatus.RESULTS_REVIEW,
        StudyStatus.FAILED_RETRYABLE,
        StudyStatus.FAILED_FINAL,
    },
    StudyStatus.RESULTS_REVIEW: {StudyStatus.PRIVACY_REVIEW},
    StudyStatus.PRIVACY_REVIEW: {
        StudyStatus.REPORT_READY,
        StudyStatus.BLOCKED_BY_POLICY,
    },
    StudyStatus.REPORT_READY: {StudyStatus.ARCHIVED},
    StudyStatus.FAILED_RETRYABLE: {
        StudyStatus.FEASIBILITY_RUNNING,
        StudyStatus.TRAINING_RUNNING,
    },
}


def transition(current: StudyStatus, target: StudyStatus) -> StudyStatus:
    if target not in ALLOWED_TRANSITIONS.get(current, set()):
        raise InvalidTransition(f"Cannot transition study from {current} to {target}")
    return target
